File: Lab2/src/ClassArray.java
import java.io.Serializable;
import java.util.Dictionary;
import java.util.Hashtable;
import java.util.Iterator;

public class ClassArray<T extends ITransport>
		implements Serializable, Iterable<T>, Iterator<T>, Comparable<ClassArray<T>> {

	private Dictionary<Integer, T> places;
	int maxCount;
	private T defaultValue;

	public ClassArray(int size, T defVal) {
		defaultValue = defVal;
		places = new Hashtable<Integer, T>();
		maxCount = size;
	}

	public static <T extends ITransport> int plus(ClassArray<T> p, T plane)
			throws AerodromeOverflowException, AerodromeAlreadyHaveException {
		boolean isFigther = plane instanceof Figther;
		if (p.places.size() == p.maxCount)
			throw new AerodromeOverflowException();
		int index = p.places.size();
		for (int i = 0; i < p.places.size(); i++) {
			if (p.checkFree(i))
				index = i;
			if (plane.getClass() == p.places.get(i).getClass()) {
				if (isFigther) {
					if (((Figther) plane).equals(p.places.get(i))) {
						throw new AerodromeAlreadyHaveException();
					}
				} else if (((War) plane).equals(p.places.get(i))) {
					throw new AerodromeAlreadyHaveException();
				}
			}
		}
		if (index != p.places.size()) {
			p.places.put(index, plane);
			return index;
		}
		p.places.put(p.places.size(), plane);
		return p.places.size() - 1;
	}

	public static <T extends ITransport> T minus(ClassArray<T> p, int index) throws AerodromeIndexOutOfRangeException {
		if (p.places.get(index) != null) {
			T plane = p.places.get(index);
			p.places.remove(index);
			return plane;
		}
		throw new AerodromeIndexOutOfRangeException();
	}

	public boolean checkFree(int index) {
		if (places.get(index) == null)
			return true;
		return false;
	}

	public T getPlane(int ind) {
		if (places.get(ind) != null)
			return places.get(ind);
		return defaultValue;
	}

	@Override
	public Iterator<T> iterator() {
		return this;
	}

	private int currentIndex;

	@Override
	public boolean hasNext() {
		if (currentIndex + 1 >= places.size()) {
			currentIndex = -1;
			return false;
		}
		currentIndex++;
		return true;
	}

	@Override
	public T next() {
		return (T) places.get(currentIndex);
	}

	@Override
	public int compareTo(ClassArray<T> other) {
		if (places.size() > other.places.size())
			return -1;
		else if (places.size() < other.places.size())
			return 1;
		else {
			for (int i = 0; i < places.size(); i++) {
				if ((places.get(places.keys().nextElement()) instanceof Figther)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Figther)) {
					return (((Figther) places.get(places.keys().nextElement()))
							.compareTo(((Figther) other.places.get(other.places.keys().nextElement()))));
				}
				if ((places.get(places.keys().nextElement()) instanceof War)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Figther))
					return 1;
				if ((places.get(places.keys().nextElement()) instanceof Figther)
						&& (other.places.get(other.places.keys().nextElement()) instanceof War))
					return -1;
				if ((places.get(places.keys().nextElement()) instanceof War)
						&& (other.places.get(other.places.keys().nextElement()) instanceof War)) {
					return (((War) places.get(places.keys().nextElement()))
							.compareTo(((War) other.places.get(other.places.keys().nextElement()))));
				}

			}
		}
		return 0;
	}

}

File: Lab 5/Laba4/Lab2/ClassArray.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab2
{
    class ClassArray<T>: IEnumerator<T>,IEnumerable<T>,IComparable<ClassArray<T>> where T : ITransport
    {

        private Dictionary<int, T> places;
        int maxCount;
        private T defaultValue;
        private int currentIndex;

        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            places = new Dictionary<int, T>();
            maxCount = size;
        }

        public static int operator +(ClassArray<T> p, T plane)
        {
            var isFighter = plane is Fighter;
            if (p.places.Count == p.maxCount) throw new AerodromeOverflowException();
            int index = p.places.Count;
            for (int i = 0; i < p.places.Count; i++)
            {
                if (p.checkFree(i))
                {
                    index = i;
                }
                if (plane.GetType() == p.places[i].GetType())
                {
                    if (isFighter)
                    {
                        if ((plane as Fighter).Equals(p.places[i])) throw new AerodromeAlreadyHaveException();
                    }
                    else if ((plane as War).Equals(p.places[i])) throw new AerodromeAlreadyHaveException();
                }
            }
            if (index != p.places.Count)
            {
                p.places.Add(index, plane);
                return index;
            }
            p.places.Add(p.places.Count, plane);
            return p.places.Count - 1;
        }

        public static T operator -(ClassArray<T> p, int index)
        {
            if (p.places.ContainsKey(index))
            {
                T plane = p.places[index];
                p.places.Remove(index);
                return plane;
            }
            throw new AerodromeIndexOutOfRangeException();
        }

        public bool checkFree(int index)
        {
            return !places.ContainsKey(index);
        }

        public T this[int ind]
        {
            get
            {
                if (places.ContainsKey(ind)) return places[ind];
                return defaultValue;
            }

        }

        public T Current
        {
            get
            {
                return places[places.Keys.ToList()[currentIndex]];
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return Current;
            }
        }

        public void Dispose() { }

        public bool MoveNext()
        {
            if (currentIndex + 1 >= places.Count)
            {
                Reset();
                return false;
            }
            currentIndex++;
            return true;
        }

        public void Reset()
        {
            currentIndex = -1;
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public int CompareTo(ClassArray<T> other)
        {
            if (this.Count() > other.Count()) return -1;
            else if (this.Count() < other.Count()) return 1;
            else
            {
                var thisKeys = this.places.Keys.ToList();
                var otherKeys = other.places.Keys.ToList();
                for(int i = 0; i < this.places.Count; ++i)
                {
                    if (this.places[thisKeys[i]] is Fighter && other.places[otherKeys[i]] is Fighter) return (this.places[thisKeys[i]] as Fighter).CompareTo(other.places[otherKeys[i]] as Fighter);
                    if (this.places[thisKeys[i]] is Fighter && other.places[otherKeys[i]] is War) return -1;
                    if (this.places[thisKeys[i]] is War && other.places[otherKeys[i]] is Fighter) return 1;
                    if (this.places[thisKeys[i]] is War && other.places[otherKeys[i]] is War) return (this.places[thisKeys[i]] as War).CompareTo(other.places[otherKeys[i]] as War);
                    
                }
            }
            return 0;
        }

    }
}
